src/VisualStudio/IntegrationTests/CSharp/CSharpAutomaticBraceCompletion.cs

```diff
@@ -1,208 +1,523 @@
         }
 
-        public void Dispose()
+        [Fact, Trait(Traits.Feature, Traits.Features.AutomaticCompletion)]
+        public void Braces_Overtyping()
         {
-            _visualStudio.Dispose();
+            SetUpEditor(@"
+class C {
+    void Foo() {
+        $$
+    }
+}");
+
+            SendKeys("if (true) {");
+            VerifyCurrentLineText("if (true) { $$}");
+
+            SendKeys("}");
+            VerifyCurrentLineText("if (true) { }$$");
         }
 
-        [Fact]
-        public async Task BracesInsertionAndTabCompleting()
+        [Fact, Trait(Traits.Feature, Traits.Features.AutomaticCompletion)]
+        public void Braces_OnReturnNoFormattingOnlyIndentationBeforeCloseBrace()
         {
-            _editorWindow.Text = @"class C {
+            SetUpEditor(@"
+class C {
     void Foo() {
-        // Marker
+        $$
     }
-}";
-
-            _editorWindow.PlaceCursor("// Marker");
+}");
 
-            await _editorWindow.TypeTextAsync("if (true) {");
-            
-            Assert.Equal("        if (true) { ", _editorWindow.CurrentLineTextBeforeCursor);
-            Assert.Equal("}", _editorWindow.CurrentLineTextAfterCursor);
+            SendKeys(
+                "if (true) {",
+                VirtualKey.Enter,
+                "var a = 1;");
 
-            await _editorWindow.TypeTextAsync($"{EditorWindow.TAB}");
-
-            Assert.Equal("        if (true) { }", _editorWindow.CurrentLineTextBeforeCursor);
-            Assert.Equal(string.Empty, _editorWindow.CurrentLineTextAfterCursor);
+            VerifyTextContains(@"
+class C {
+    void Foo() {
+        if (true)
+        {
+            var a = 1;$$
+        }
+    }
+}");
         }
 
-        [Fact]
-        public async Task BracesOvertyping()
+        [Fact, Trait(Traits.Feature, Traits.Features.AutomaticCompletion)]
+        public void Braces_OnReturnOvertypingTheClosingBrace()
         {
-            _editorWindow.Text = @"class C {
+            SetUpEditor(@"
+class C {
     void Foo() {
-        // Marker
+        $$
     }
-}";
+}");
 
-            _editorWindow.PlaceCursor("// Marker");
+            SendKeys(
+                "if (true) {",
+                VirtualKey.Enter,
+                "var a = 1;",
+                '}');
 
-            await _editorWindow.TypeTextAsync("if (true) {");
-            await _editorWindow.TypeTextAsync("}");
+            VerifyTextContains(@"
+class C {
+    void Foo() {
+        if (true)
+        {
+            var a = 1;
+        }$$
+    }
+}");
+        }
 
-            Assert.Equal("        if (true) { }", _editorWindow.CurrentLineTextBeforeCursor);
-            Assert.Equal(string.Empty, _editorWindow.CurrentLineTextAfterCursor);
+        [WorkItem(653540, "DevDiv")]
+        [Fact, Trait(Traits.Feature, Traits.Features.AutomaticCompletion)]
+        public void Braces_OnReturnWithNonWhitespaceSpanInside()
+        {
+            SetUpEditor("$$");
```

Comment:
This is done. I also added a special `ClearEditor` API to handle this operation for tests that need to use it.